### PR TITLE
Delay support when loading from QASM2

### LINKED ISSUE #6510: cannot load qasms containing delay
e.g.

```c
OPENQASM 2.0;
include "qelib1.inc";
qreg q[4];
creg meas[4];
delay(172) q[1];
u3(1.955526,-3.9994262,1.5396532) q[1];
```

### PR DESCRIPTION
**Pull request by 1ucian0**
Fixes #6510

This PR maps the string `delay` to the `qiskit.circuit.delay.Delay` object.

```
qasm = """OPENQASM 2.0;
include "qelib1.inc";

opaque delay(time) q;

qreg q[1];

delay(172) q[0];
u3(0.1,0.2,0.3) q[0];
"""
circuit = QuantumCircuit.from_qasm_str(qasm)
circuit.draw()
```
```
     ┌────────────────┐┌─────────────────┐
q_0: ┤ DELAY(172[dt]) ├┤ U3(0.1,0.2,0.3) ├
     └────────────────┘└─────────────────┘
```

### CODE REVIEW

File: qiskit/circuit/delay.py [hunk 1]
```diff
@@ -72,7 +72,10 @@ def validate_parameter(self, parameter):
             return parameter
         elif isinstance(parameter, float):
             if self.unit == "dt":
-                raise CircuitError("Integer duration is expected for 'dt' unit.")
+                parameter_int = int(parameter)
```
Comment (kdk) [inline qiskit/circuit/delay.py:75]: Is this needed for loading from QASM? (Does the QASM parser parse all numbers a floats?)

**Review by kdk — COMMENTED** [review]
This looks good so far. Will this handle the case where `unit` is something other than `'dt'`?

File: qiskit/circuit/delay.py (hunk 1 above)
Comment (1ucian0) [reply qiskit/circuit/delay.py:75]: QASM parser converts all the non-qubits paramters into floats, yes.

**Comment by 1ucian0** [comment]
> This looks good so far. Will this handle the case where `unit` is something other than `'dt'`?

Nope :( because in the specification says that all the *real parameters* are, well, real numbers (same reason [why the parser convert them to floats](https://github.com/Qiskit/qiskit-terra/pull/6594#discussion_r654537709)):

Screenshot 2021-06-18 at 18 18 47

Screenshot 2021-06-18 at 18 20 54

The goal of `opaque` was to cover future operations. But ended up very limiting, like calling it "gates" (which is not the current case) and only allow real parameters.

Alternative, the option is to consider `delay` an annotation. But for that, [qiskit need to parse comments](https://github.com/Qiskit/qiskit-terra/issues/901).

**Comment by kdk** [comment]
> The goal of `opaque` was to cover future operations. But ended up being very limited, like calling it "gates" (which is not the current case) and only allow real parameters.
> 
> Alternative, the option is to consider `delay` an annotation. But for that, [qiskit need to parse comments](https://github.com/Qiskit/qiskit-terra/issues/901).

That makes sense. It'd be good though to avoid losing unit information on a export -> parse though, if we can. In the worst case, can we warn or raise on export if `self.unit != 'dt'`? Or maybe export and parse as `delay_dt`, `delay_ps`, `delay_ns`, ... ?

**Comment by 1ucian0** [comment]
In addition `delay_dt`, `delay_ps`, and `delay_ns` in addition to `delay`? I'm taking the default right now.

**Comment by kdk** [comment]
> `delay_dt`, `delay_ps`, and `delay_ns` in addition to `delay`? I'm taking the default unit right now.
> 
> > can we warn or raise on export if ...
> 
> Just to be sure, this PR is just about parsing, not exporting.

Fair point, opened this as a new issue ( #6643 ).

**Review by kdk — APPROVED** [review]

**Pull request merged**